pennylane_lightning/src/simulator/StateVectorBase.hpp

```diff
@@ -0,0 +1,667 @@
+// Copyright 2021 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+/**
+ * @file
+ * Defines the class representation for quantum state vectors.
+ */
+
+#pragma once
+
+#include "DynamicDispatcher.hpp"
+#include "Error.hpp"
+#include "Gates.hpp"
+#include "SelectGateOps.hpp"
+#include "Util.hpp"
+
+/// @cond DEV
+// Required for compilation with MSVC
+#ifndef _USE_MATH_DEFINES
+#define _USE_MATH_DEFINES // for C++
+#endif
+/// @endcond
+
+#include <cmath>
+#include <complex>
+#include <functional>
+#include <iostream>
+#include <set>
+#include <stdexcept>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+/**
+ * @brief This macro defines methods for State-vector class. The kernel template
+ * argument choose the kernel to run.
+ */
+#define PENNYLANE_STATEVECTOR_DEFINE_OPS(GATE_NAME)                            \
```

Comment:
While I know macros work here, I wonder if there is a better way to do this with some template metaprogramming that avoids the preprocessor altogether. 

I think for now, that may be a challenge, but something akin to https://stackoverflow.com/a/15444681 could be useful as a follow up. By avoiding the preprocessor, all generated code and options are only worked by the C++ engine, and we can do some nice static-analysis on it as needed.

Reply:
Good idea. I looked for some different workarounds but they both seem to need a huge changes in interface. As I want to call the function directly with `sv.applyPauliX(...)`, I definitely need to define a method. And currently, I don't think TMP can be used to define a method. A possible solution is introduced in https://www.youtube.com/watch?v=8c6BAQcYF_E&t=2596s, but it needs reflection TS (which is unlikely to be included even in C++23).

Reply:
Good point -- yea the reflection TS looks beneficial. I guess we wait until C++2z support so. Happy to keep this in that case.